### ISSUE TITLE
Possible integer overflow in GEMV

### ISSUE THREAD

**Issue by twesterhout** [primme/primme#43]
I just ran into an issue that `dprimme` crashed with the following error:
```
 ** On entry to DGEMV  parameter number  6 had an illegal value
```
PRIMME is using Fortran interface to BLAS, so parameter number 6 is `lda` in
```c
dgemv (trans, m, n, alpha, a, lda, x, incx, beta, y, incy)
```

Dimension of the problem is 3204236779 which is more than can fit into a 32-bit `int` (PRIMME was compiled with `PRIMME_INT_SIZE=64` & `PRIMME_BLASINT_SIZE=32`). Looking at the code:

https://github.com/primme/primme/blob/60055676dfa4f9c4fc9824dfa208b83cbd8b4e5b/src/linalg/blaslapack.c#L747

`gemv` is applied multiple times and care is taken to properly split `m` into smaller chunks, but `lda` can still overflow.

Am I missing something or is this a bug?

Thanks!

**Comment by eromero-vlc** [comment]
Yes, that's a bug. That code is stupid and just should report an error.

The easiest workaround is to use more mpi processes so that each process has to deal with a small enough number of rows to be handle by blas. The other alternative is to use a BLAS that supports 64bit int, like MKL and openblas, and recompile primme with PRIMME_INT_SIZE=64 and PRIMME_BLASINT_SIZE=64.

**Comment by twesterhout** [comment]
I'll try switching to 64-bit BLAS, thanks!

**Comment by eromero-vlc** [comment]
Thanks for the report! I'll fixed it!

**Comment by twesterhout** [comment]
Okay, `PRIMME_BLASINT_SIZE=64` is not cutting it because `Num_gemv_Sprimme` declares `lda` as `int` and the overflow happens before calling into BLAS.

**Comment by eromero-vlc** [comment]
I have uploaded a patch to master that fixed that. I tried to run it with openblas compiled with 64 bit int, but I think that openblas does not truly allow blas calls with large dimensions. I have to tested with mkl.

**Comment by twesterhout** [comment]
Ah this is great, thank you! I'll give it a try and report back.

**Comment by twesterhout** [comment]
First tests seem successful. It's not crashing anymore (with Intel MKL 2019.5)!

**Comment by eromero-vlc** [comment]
Happy to hear that. Let us know if you find more issues.

**Comment by eromero-vlc** [comment]
BTW I am not sure if I uploaded the patch when I said I did 13 days ago. If you continue experiencing problems, update master branch and let us know if you have issues.

**Comment by twesterhout** [comment]
Yes, you did, but to `eloy/fix-large-mats` branch rather than `master`. I've [been building](https://github.com/twesterhout/primme-hs/blob/master/configure#L43) from that branch, but I guess now I can switch to `master`.

**Comment by eromero-vlc** [comment]
OMG! You built an interface in Haskell! I love it! Keep it up!
